teamcity: failed tests on master: acceptance/TestGossipRestart

The following tests appear to have failed:

[#573171](https://teamcity.cockroachdb.com/viewLog.html?buildId=573171):


```
--- FAIL: acceptance/TestGossipRestart: TestGossipRestartFirstNodeNeedsIncoming (8.740s)
test_log_scope.go:81: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestGossipRestartFirstNodeNeedsIncoming566027454
test_log_scope.go:62: use -show-logs to present logs inline
--- FAIL: acceptance/TestGossipRestart: TestGossipRestartFirstNodeNeedsIncoming: TestGossipRestartFirstNodeNeedsIncoming/runMode=local (8.740s)
gossip_peerings_test.go:246: cmd: [/home/agent/work/.go/src/github.com/cockroachdb/cockroach/cockroach zone set .default -f /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/.customzone --insecure --port 45913]
	stderr:
	 
	stdout:
	 Error: pq: restart transaction: HandledRetryableTxnError: TransactionAbortedError: txn aborted "sql txn" id=5e93e827 key=/Table/SystemConfigSpan/Start rw=true pri=0.06785155 iso=SERIALIZABLE stat=PENDING epo=0 ts=1520885073.552174506,0 orig=1520885073.552174506,0 max=1520885073.552174506,0 wto=false rop=false seq=10
	Failed running "zone"
	: exit status 1
------- Stdout: -------
CockroachDB node starting at 2018-03-12 20:04:29.384600847 +0000 UTC (took 0.5s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:36401
sql:                 postgresql://root@127.0.0.1:33429?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/1
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/cockroach-temp610578941
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1
status:              initialized new cluster
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              1
CockroachDB node starting at 2018-03-12 20:04:30.073003063 +0000 UTC (took 0.2s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:36817
sql:                 postgresql://root@127.0.0.1:34065?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/3
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/3/cockroach-temp951990825
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/3/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/3
status:              initialized new node, joined pre-existing cluster
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              2
CockroachDB node starting at 2018-03-12 20:04:30.079848273 +0000 UTC (took 0.2s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:38443
sql:                 postgresql://root@127.0.0.1:45373?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/2
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/2/cockroach-temp448940807
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/2/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/2
status:              initialized new node, joined pre-existing cluster
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              3
CockroachDB node starting at 2018-03-12 20:04:30.086290771 +0000 UTC (took 0.2s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:44585
sql:                 postgresql://root@127.0.0.1:46277?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/4
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/4/cockroach-temp843042692
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/4/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/4
status:              initialized new node, joined pre-existing cluster
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              4
CockroachDB node starting at 2018-03-12 20:04:33.102937058 +0000 UTC (took 0.7s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:36059
sql:                 postgresql://root@127.0.0.1:45913?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/1
attrs:               empty
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/cockroach-temp930306299
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1
status:              restarted pre-existing node
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              1
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestGossipRestartFirstNodeNeedsIncoming566027454
--- FAIL: acceptance/TestGossipRestart: TestGossipRestartFirstNodeNeedsIncoming (8.740s)
test_log_scope.go:81: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestGossipRestartFirstNodeNeedsIncoming566027454
test_log_scope.go:62: use -show-logs to present logs inline
--- FAIL: acceptance/TestGossipRestart: TestGossipRestartFirstNodeNeedsIncoming: TestGossipRestartFirstNodeNeedsIncoming/runMode=local (8.740s)
gossip_peerings_test.go:246: cmd: [/home/agent/work/.go/src/github.com/cockroachdb/cockroach/cockroach zone set .default -f /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/.customzone --insecure --port 45913]
	stderr:
	 
	stdout:
	 Error: pq: restart transaction: HandledRetryableTxnError: TransactionAbortedError: txn aborted "sql txn" id=5e93e827 key=/Table/SystemConfigSpan/Start rw=true pri=0.06785155 iso=SERIALIZABLE stat=PENDING epo=0 ts=1520885073.552174506,0 orig=1520885073.552174506,0 max=1520885073.552174506,0 wto=false rop=false seq=10
	Failed running "zone"
	: exit status 1
------- Stdout: -------
CockroachDB node starting at 2018-03-12 20:04:29.384600847 +0000 UTC (took 0.5s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:36401
sql:                 postgresql://root@127.0.0.1:33429?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/1
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/cockroach-temp610578941
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1
status:              initialized new cluster
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              1
CockroachDB node starting at 2018-03-12 20:04:30.073003063 +0000 UTC (took 0.2s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:36817
sql:                 postgresql://root@127.0.0.1:34065?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/3
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/3/cockroach-temp951990825
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/3/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/3
status:              initialized new node, joined pre-existing cluster
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              2
CockroachDB node starting at 2018-03-12 20:04:30.079848273 +0000 UTC (took 0.2s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:38443
sql:                 postgresql://root@127.0.0.1:45373?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/2
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/2/cockroach-temp448940807
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/2/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/2
status:              initialized new node, joined pre-existing cluster
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              3
CockroachDB node starting at 2018-03-12 20:04:30.086290771 +0000 UTC (took 0.2s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:44585
sql:                 postgresql://root@127.0.0.1:46277?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/4
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/4/cockroach-temp843042692
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/4/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/4
status:              initialized new node, joined pre-existing cluster
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              4
CockroachDB node starting at 2018-03-12 20:04:33.102937058 +0000 UTC (took 0.7s)
build:               CCL v2.0-alpha.20180212-1504-gc438631 @ 2018/03/12 19:54:55 (go1.10)
admin:               http://127.0.0.1:36059
sql:                 postgresql://root@127.0.0.1:45913?application_name=cockroach&sslmode=disable
logs:                /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestGossipRestartFirstNodeNeedsIncoming/runMode=local/1
attrs:               empty
temp dir:            /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/cockroach-temp930306299
external I/O path:   /home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1/extern
store[0]:            path=/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster734920453/1
status:              restarted pre-existing node
clusterID:           1b55cfdb-80fa-4f1e-b326-891c1c8c72c2
nodeID:              1
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestGossipRestartFirstNodeNeedsIncoming566027454

```

Please assign, take a look and update the issue accordingly.

--- a/pkg/cli/sql_util.go
+++ b/pkg/cli/sql_util.go
@@ -17,6 +17,7 @@ package cli
 import (
 	"bytes"
 	"context"
+	gosql "database/sql"
 	"database/sql/driver"
 	"fmt"
 	"io"
@@ -27,13 +28,13 @@ import (
 	"unicode"
 	"unicode/utf8"
 
+	"github.com/cockroachdb/cockroach-go/crdb"
 	version "github.com/hashicorp/go-version"
 	"github.com/lib/pq"
 	"github.com/pkg/errors"
 
 	"github.com/cockroachdb/cockroach/pkg/build"
 	"github.com/cockroachdb/cockroach/pkg/security"
-	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -264,58 +265,45 @@ func (c *sqlConn) getServerValue(what, sql string) (driver.Value, bool) {
 	return dbVals[0], true
 }
 
+// sqlTxnShim implements the crdb.Tx interface.
+//
+// It exists to support crdb.ExecuteInTxn. Normally, we'd hand crdb.ExecuteInTxn
+// a sql.Txn, but sqlConn predates go1.8's support for multiple result sets and
+// so deals directly with the lib/pq driver. See #14964.
+type sqlTxnShim struct {
+	conn *sqlConn
+}
+
+func (t sqlTxnShim) Commit() error {
+	return t.conn.Exec(`COMMIT`, nil)
+}
+
+func (t sqlTxnShim) Rollback() error {
+	return t.conn.Exec(`ROLLBACK`, nil)
+}
+
+func (t sqlTxnShim) ExecContext(
+	_ context.Context, query string, values ...interface{},
+) (gosql.Result, error) {
+	if len(values) != 0 {
+		panic(fmt.Sprintf("sqlTxnShim.ExecContext must not be called with values"))
+	}
+	return nil, t.conn.Exec(query, nil)
+}
+
 // ExecTxn runs fn inside a transaction and retries it as needed.
 // On non-retryable failures, the transaction is aborted and rolled
 // back; on success, the transaction is committed.
 //
 // NOTE: the supplied closure should not have external side
 // effects beyond changes to the database.
-//
-// NB: this code is cribbed from cockroach-go/crdb and has been copied
-// because this code, pre-dating go1.8, deals with multiple result sets
-// direct with the driver. See #14964.
 func (c *sqlConn) ExecTxn(fn func(*sqlConn) error) (err error) {
-	// Start a transaction.
-	if err = c.Exec(`BEGIN`, nil); err != nil {
+	if err := c.Exec(`BEGIN`, nil); err != nil {
 		return err
 	}
-	defer func() {
-		if err == nil {
-			// Ignore commit errors. The tx has already been committed by RELEASE.
-			_ = c.Exec(`COMMIT`, nil)
-		} else {
-			// We always need to execute a Rollback() so sql.DB releases the
-			// connection.
-			_ = c.Exec(`ROLLBACK`, nil)
-		}
-	}()
-	// Specify that we intend to retry this txn in case of CockroachDB retryable
-	// errors.
-	if err = c.Exec(`SAVEPOINT cockroach_restart`, nil); err != nil {
-		return err
-	}
-
-	for {
-		err = fn(c)
-		if err == nil {
-			// RELEASE acts like COMMIT in CockroachDB. We use it since it gives us an
-			// opportunity to react to retryable errors, whereas tx.Commit() doesn't.
-			if err = c.Exec(`RELEASE SAVEPOINT cockroach_restart`, nil); err == nil {
-				return nil
-			}
-		}
-		// We got an error; let's see if it's a retryable one and, if so, restart. We look
-		// for either the standard PG errcode SerializationFailureError:40001 or the Cockroach extension
-		// errcode RetriableError:CR000. The Cockroach extension has been removed server-side, but support
-		// for it has been left here for now to maintain backwards compatibility.
-		pqErr, ok := pgerror.GetPGCause(err)
-		if retryable := ok && (pqErr.Code == "CR000" || pqErr.Code == "40001"); !retryable {
-			return err
-		}
-		if err = c.Exec(`ROLLBACK TO SAVEPOINT cockroach_restart`, nil); err != nil {
-			return err
-		}
-	}
+	return crdb.ExecuteInTx(context.TODO(), sqlTxnShim{c}, func() error {
+		return fn(c)
+	})
 }
 
 func (c *sqlConn) Exec(query string, args []driver.Value) error {

--- a/pkg/cli/sql_util_test.go
+++ b/pkg/cli/sql_util_test.go
@@ -207,3 +207,47 @@ SET
 	}
 	b.Reset()
 }
+
+func TestTransactionRetry(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	p := cliTestParams{t: t}
+	c := newCLITest(p)
+	defer c.cleanup()
+
+	url, cleanup := sqlutils.PGUrl(t, c.ServingAddr(), t.Name(), url.User(security.RootUser))
+	defer cleanup()
+
+	conn := makeSQLConn(url.String())
+	defer conn.Close()
+
+	var tries int
+	err := conn.ExecTxn(func(conn *sqlConn) error {
+		tries++
+		if tries > 2 {
+			return nil
+		}
+
+		// Prevent automatic server-side retries.
+		rows, err := conn.Query(`SELECT now()`, nil)
+		if err != nil {
+			return err
+		}
+		if err := rows.Close(); err != nil {
+			return err
+		}
+
+		// Force a client-side retry.
+		rows, err = conn.Query(`SELECT crdb_internal.force_retry('1h')`, nil)
+		if err != nil {
+			return err
+		}
+		return rows.Close()
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tries <= 2 {
+		t.Fatalf("expected transaction to require at least two tries, but it only required %d", tries)
+	}
+}

